State Timeline: Fix crash when hovering over panel

hugohaggmark opened the pull request:
**What this PR does / why we need it**:
This is a new area in the code base for me so I don't fully understand why we were using a lookup to access a field that we already have, I'm certain there were good reasons and I'll change this fix accordingly. This fix replaces the lookup based approach to just passing the field to `getFieldDisplayName`.

Use this dashboard to repro the issue on main:
```json

{
  "__inputs": [],
  "__requires": [
    {
      "type": "grafana",
      "id": "grafana",
      "name": "Grafana",
      "version": "8.1.0-pre"
    },
    {
      "type": "panel",
      "id": "state-timeline",
      "name": "State timeline",
      "version": ""
    }
  ],
  "annotations": {
    "list": [
      {
        "builtIn": 1,
        "datasource": "-- Grafana --",
        "enable": true,
        "hide": true,
        "iconColor": "rgba(0, 211, 255, 1)",
        "name": "Annotations & Alerts",
        "type": "dashboard"
      }
    ]
  },
  "editable": true,
  "gnetId": null,
  "graphTooltip": 0,
  "id": null,
  "links": [],
  "panels": [
    {
      "datasource": null,
      "fieldConfig": {
        "defaults": {
          "color": {
            "mode": "thresholds"
          },
          "custom": {
            "fillOpacity": 70,
            "lineWidth": 0
          },
          "mappings": [],
          "thresholds": {
            "mode": "absolute",
            "steps": [
              {
                "color": "green",
                "value": null
              },
              {
                "color": "red",
                "value": 80
              }
            ]
          }
        },
        "overrides": []
      },
      "gridPos": {
        "h": 9,
        "w": 12,
        "x": 0,
        "y": 0
      },
      "id": 2,
      "options": {
        "alignValue": "left",
        "legend": {
          "displayMode": "list",
          "placement": "bottom"
        },
        "mergeValues": true,
        "rowHeight": 0.9,
        "showValue": "auto",
        "tooltip": {
          "mode": "single"
        }
      },
      "targets": [
        {
          "csvWave": [
            {
              "labels": "",
              "name": "kalle",
              "timeStep": 182,
              "valuesCSV": "0,1,100"
            },
            {
              "timeStep": 60,
              "valuesCSV": "0,0,92,2,1,1"
            }
          ],
          "refId": "A",
          "scenarioId": "predictable_csv_wave"
        }
      ],
      "title": "Panel Title",
      "transformations": [
        {
          "id": "filterFieldsByName",
          "options": {
            "include": {
              "names": [
                "Time",
                "A-series"
              ]
            }
          }
        }
      ],
      "type": "state-timeline"
    }
  ],
  "schemaVersion": 30,
  "style": "dark",
  "tags": [],
  "templating": {
    "list": []
  },
  "time": {
    "from": "now-6h",
    "to": "now"
  },
  "timepicker": {},
  "timezone": "",
  "title": "State Timeline with Filter by Name Transform crashes #35601",
  "uid": "AUiqwDg7k",
  "version": 4
}

```

**Which issue(s) this PR fixes**:
Fixes #35601
Fixes #35752

**Special notes for your reviewer**:

hugohaggmark committed:
State Timeline: Fix crash when hovering over panel

thisisobate reviewed (approved):
LGTM! But it seems trying to zoom in a visualization- in this case State Timeline- causes Chrome to freeze. Maybe we can open this as a separate issue on GH @hugohaggmark

dprokop commented:
We use the lookup access to make sure we are always performing operations on the same data structure - the original data frame that was provided to the panel. 

Let me explain the process:
1. Panel gets DFs that get passed down to uPlot based visualization
2. In the particular viz component (i.e. Timeline) the data frames get prepared for visualization and then passed down to GraphNG
3. GraphNG **prepares the DFs for uPlot**, by default outer joining them by provided fields (time field by default): https://github.com/grafana/grafana/blob/main/packages/grafana-ui/src/components/GraphNG/GraphNG.tsx#L89 - the default implementation provides te origin indices. 

The data alignment/preparation can have different implementations making changes to the original fields(idealy copying them prior aplying the change). The index lookup makes sure that the names are calculated always on the "raw" data.

Hope this helps a bit understanding the flow @hugohaggmark .

In this particular case, since the field you are using is not altered anyhow during the process, we can rely on it. And the fix should work OK.

But after debugging the code it feels to me like the indices are wrong in the Timeline panel's aligned frame, as the index points to not a field that was never there:

![image](https://user-images.githubusercontent.com/2376619/122050525-d1704c00-cde3-11eb-9aad-e7aaa4ed20e5.png)
![image](https://user-images.githubusercontent.com/2376619/122050730-0da3ac80-cde4-11eb-8bac-9c9e57c1896d.png)

Would probably dig into this a bit too.

dprokop commented:
@hugohaggmark actually I think we are making a mistake during fields preparation for the timeline panel (https://github.com/grafana/grafana/blob/b774dd9b1a229c35a8d16d0f21bf281f713623a8/public/app/plugins/panel/state-timeline/utils.ts#L258) where we assume that time field presence indicates timeseries, while it should be both time and value.

hugohaggmark commented:
@dprokop I think somewhere the `field.state.origin` is pointing to the wrong frameIndex but I wasn't sure where. Not sure how to proceed as I'm not familiar with the inner works of GraphNG.

hugohaggmark commented:
> LGTM! But it seems trying to zoom in a visualization- in this case State Timeline- causes Chrome to freeze. Maybe we can open this as a separate issue on GH @hugohaggmark

@thisisobate can you check if this issue is already in `main` or if this PR has introduced it. If it's already in main can you create a new issue with good repro steps? Thank you!

dprokop commented:
@hugohaggmark take a look at my comment above about the `prepareTimelineFields` function. As long as you make sure the returned DFs are graphable (have time + other fields(must)) then this should fix the issue

dprokop commented:
All I wrote is wrong:D StateTimelineTooltip gets wrong data: data.series instead of prepared frames (`frames`). Sorry!

hugohaggmark committed:
Refactor: fixes after PR comments

hugohaggmark committed:
Chore: fixes infinite loop

hugohaggmark commented:
Ok, updated according to @dprokop suggestion, thank you for helping out with that. And I also fixed an infinite loop in the utils when zooming in the time line panel.

thisisobate reviewed (approved):
LGTM!
